src/wasm/wasm-emscripten.cpp

```diff
@@ -702,45 +717,110 @@ void AsmConstWalker::noteNonLinear(Expression* curr) {
 
 void AsmConstWalker::visitLocalSet(LocalSet* curr) { sets[curr->index] = curr; }
 
+Const* AsmConstWalker::resolveConstAddr(Expression* arg, const Name& target) {
+  while (!arg->dynCast<Const>()) {
+    if (auto* get = arg->dynCast<LocalGet>()) {
+      // The argument may be a local.get, in which case, the last set in this
+      // basic block has the value.
+      auto* set = sets[get->index];
+      if (set) {
+        assert(set->index == get->index);
+        arg = set->value;
+      }
+    } else if (auto* value = arg->dynCast<Binary>()) {
+      // In the dynamic linking case the address of the string constant
+      // is the result of adding its offset to __memory_base.
+      // In this case are only looking for the offset with the data segment so
+      // the RHS of the addition is just what we want.
+      assert(value->op == AddInt32);
+      arg = value->right;
+    } else {
+      Fatal() << "Unexpected arg0 type (" << getExpressionName(arg)
+              << ") in call to to: " << target;
+    }
+  }
+  return arg->cast<Const>();
+}
+
+Index AsmConstWalker::resolveConstIndex(
+  Expression* arg, std::function<void(Expression*)> reportError) {
+  while (!arg->dynCast<Const>()) {
+    if (auto* get = arg->dynCast<LocalGet>()) {
+      // The argument may be a local.get, in which case, the last set in this
+      // basic block has the value.
+      auto* set = sets[get->index];
+      if (set) {
+        assert(set->index == get->index);
+        arg = set->value;
+      } else {
+        reportError(arg);
+        return 0;
+      }
+    } else if (arg->is<GlobalGet>()) {
+      // In the dynamic linking case, indices start at __table_base.
+      // We want the value relative to __table_base.
+      return 0;
```

Comment:
It is unclear to me how returning 0 fulfills the requirements stated in the comment. Perhaps the comment can be clarified?